### ISSUE TITLE
Update podspec?

### ISSUE THREAD

**Issue by robertmryan** [ccgus/fmdb#793]
I published version 2.7.7 release (making it available for Carthage users), but when I went to push the Cocoapods trunk, I got a whole bunch of warnings:

```none
Updating spec repo `trunk`
Validating podspec
nil versions are discouraged and will be deprecated in Rubygems 4
 -> FMDB (2.7.7)
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Using new build system
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Building targets in parallel
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Planning build
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Constructing build description
    - WARN  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabaseQueue.m:101:9: warning: 'dispatch_queue_set_specific' is only available on macOS 10.7 or newer [-Wunguarded-availability]
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/dispatch/queue.h:1498:1: note: 'dispatch_queue_set_specific' has been marked as being introduced in macOS 10.7 here, but the deployment target is macOS 10.6.0
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabaseQueue.m:101:9: note: enclose 'dispatch_queue_set_specific' in an @available check to silence this warning
    - WARN  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabaseQueue.m:184:54: warning: 'dispatch_get_specific' is only available on macOS 10.7 or newer [-Wunguarded-availability]
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/dispatch/queue.h:1554:1: note: 'dispatch_get_specific' has been marked as being introduced in macOS 10.7 here, but the deployment target is macOS 10.6.0
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabaseQueue.m:184:54: note: enclose 'dispatch_get_specific' in an @available check to silence this warning
    - WARN  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:133:33: warning: 'objectAtIndexedSubscript:' is only available on macOS 10.8 or newer [-Wunguarded-availability]
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSArray.h:61:1: note: 'objectAtIndexedSubscript:' has been marked as being introduced in macOS 10.8 here, but the deployment target is macOS 10.6.0
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:133:33: note: enclose 'objectAtIndexedSubscript:' in an @available check to silence this warning
    - WARN  | [FMDB/standard,FMDB/FTS] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:1340:15: warning: 'sqlite3_wal_checkpoint_v2' is only available on macOS 10.8 or newer [-Wunguarded-availability]
    - NOTE  | [FMDB/standard,FMDB/FTS] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sqlite3.h:8428:16: note: 'sqlite3_wal_checkpoint_v2' has been marked as being introduced in macOS 10.8 here, but the deployment target is macOS 10.6.0
    - NOTE  | [FMDB/standard,FMDB/FTS] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:1340:15: note: enclose 'sqlite3_wal_checkpoint_v2' in an @available check to silence this warning
    - WARN  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:310:49: warning: 'arc4random_uniform' is only available on macOS 10.7 or newer [-Wunguarded-availability]
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/stdlib.h:273:3: note: 'arc4random_uniform' has been marked as being introduced in macOS 10.7 here, but the deployment target is macOS 10.6.0
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:310:49: note: enclose 'arc4random_uniform' in an @available check to silence this warning
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Execution policy exception registration failed and was skipped: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted" (in target 'FMDB' from project 'Pods')
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Execution policy exception registration failed and was skipped: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted" (in target 'Pods-App' from project 'Pods')
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Execution policy exception registration failed and was skipped: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted" (in target 'App' from project 'App')
    - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  warning: Skipping code signing because the target does not have an Info.plist file and one is not being generated automatically. (in target 'App' from project 'App')
    - WARN  | [FMDB/FTS,FMDB/standalone-fts] xcodebuild:  FMDB/src/extra/fts3/FMDatabase+FTS3.m:219:60: warning: 'NSPointerFunctionsWeakMemory' is only available on macOS 10.8 or newer [-Wunguarded-availability]
    - NOTE  | [FMDB/FTS,FMDB/standalone-fts] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSPointerFunctions.h:40:5: note: 'NSPointerFunctionsWeakMemory' has been marked as being introduced in macOS 10.8 here, but the deployment target is macOS 10.6.0
    - NOTE  | [FMDB/FTS,FMDB/standalone-fts] xcodebuild:  FMDB/src/extra/fts3/FMDatabase+FTS3.m:219:60: note: enclose 'NSPointerFunctionsWeakMemory' in an @available check to silence this warning
    - NOTE  | [FMDB/standalone,FMDB/standalone-fts] xcodebuild:  note: Execution policy exception registration failed and was skipped: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted" (in target 'sqlite3' from project 'Pods')
    - ERROR | [OSX] [FMDB/SQLCipher] unknown: Encountered an unknown error (CocoaPods could not find compatible versions for pod "SQLCipher":
  In Podfile:
    FMDB/SQLCipher (from `/Users/rryan/.../fmdb/FMDB.podspec`) was resolved to 2.7.7, which depends on
      SQLCipher (~> 4.0)

Specs satisfying the `SQLCipher (~> 4.0)` dependency were found, but they required a higher minimum deployment target.) during validation.

[!] The spec did not pass validation, due to 1 error and 6 warnings.
```

Most of this consists of warnings that seem like they could be fixed by specifying the correct deployment target in the podspec. But the one that’s throwing me is the last one:

```none
    - ERROR | [OSX] [FMDB/SQLCipher] unknown: Encountered an unknown error (CocoaPods could not find compatible versions for pod "SQLCipher":
  In Podfile:
    FMDB/SQLCipher (from `/Users/rryan/.../fmdb/FMDB.podspec`) was resolved to 2.7.7, which depends on
      SQLCipher (~> 4.0)
```

I’m not quite sure what to make of that as it appears to be that `SQLCipher` does have a pod for v4.4, which seems like it should match. Does anyone know what’s going on there?

(By the way, I’m not sure why it’s using `~>` rather than, say, `>=`, but that’s a separate question.)

**Comment by R4N** [comment]
>(By the way, I’m not sure why it’s using ~> rather than, say, >=, but that’s a separate question.)

Someone had raised an issue a while back that they were unintentionally migrated to the next major version of SQLCipher when using `pod update` (i.e. they were using SQLCipher 3.x and when using `pod update`, updated to SQLCipher 4.x which breaks compatibility). Using `~>` restricts it to use the greatest 4.x version of SQLCipher without going over the major version threshold. This will allow control over when to bump to the next major version of SQLCipher subspec.

>'~> 0.1' Version 0.1 and the versions up to 1.0, not including 1.0 and higher

From https://guides.cocoapods.org/using/the-podfile.html

This was the PR related with that change: https://github.com/ccgus/fmdb/pull/753 (which also links the issue)

Ideally FMDB users who use the SQLCipher subspec should have major version compatibility migration in place for SQLCipher as mentioned in this post on the SQLCipher discussion forum: https://discuss.zetetic.net/t/upgrading-to-sqlcipher-4/3283

>I’m not quite sure what to make of that as it appears to be that SQLCipher does have a pod for v4.4, which seems like it should match.

I would assume this is the case as well. I just ran through a test project which uses FMDB/SQLCipher subspec  and ran `pod update` and it properly updated to 4.4.0.

What's your minimum deployment target set to? From the error message it sounds like it found a possible version, but it's greater than the min deployment target:

>Specs satisfying the `SQLCipher (~> 4.0)` dependency were found, but they required a higher minimum deployment target.)

**Comment by R4N** [comment]
I wonder if adding these couple of lines to the FMDB podspec might help?

```
s.ios.deployment_target  = '8.0'
s.osx.deployment_target  = '10.9'
```

as those line up with the minimum specified in the SQLCipher podspec: https://github.com/sqlcipher/sqlcipher/blob/master/SQLCipher.podspec.json#L9

**Comment by tstump-phunware** [comment]
@robertmryan We should be good to close this issue out with PR #824 having been merged to master.

**Comment by tstump-phunware** [comment]
> I published version 2.7.7 release (making it available for Carthage users), but when I went to push the Cocoapods trunk, I got a whole bunch of warnings:
> 
> ```
> Updating spec repo `trunk`
> Validating podspec
> nil versions are discouraged and will be deprecated in Rubygems 4
>  -> FMDB (2.7.7)
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Using new build system
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Building targets in parallel
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Planning build
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Constructing build description
>     - WARN  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabaseQueue.m:101:9: warning: 'dispatch_queue_set_specific' is only available on macOS 10.7 or newer [-Wunguarded-availability]
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/dispatch/queue.h:1498:1: note: 'dispatch_queue_set_specific' has been marked as being introduced in macOS 10.7 here, but the deployment target is macOS 10.6.0
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabaseQueue.m:101:9: note: enclose 'dispatch_queue_set_specific' in an @available check to silence this warning
>     - WARN  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabaseQueue.m:184:54: warning: 'dispatch_get_specific' is only available on macOS 10.7 or newer [-Wunguarded-availability]
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/dispatch/queue.h:1554:1: note: 'dispatch_get_specific' has been marked as being introduced in macOS 10.7 here, but the deployment target is macOS 10.6.0
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabaseQueue.m:184:54: note: enclose 'dispatch_get_specific' in an @available check to silence this warning
>     - WARN  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:133:33: warning: 'objectAtIndexedSubscript:' is only available on macOS 10.8 or newer [-Wunguarded-availability]
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSArray.h:61:1: note: 'objectAtIndexedSubscript:' has been marked as being introduced in macOS 10.8 here, but the deployment target is macOS 10.6.0
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:133:33: note: enclose 'objectAtIndexedSubscript:' in an @available check to silence this warning
>     - WARN  | [FMDB/standard,FMDB/FTS] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:1340:15: warning: 'sqlite3_wal_checkpoint_v2' is only available on macOS 10.8 or newer [-Wunguarded-availability]
>     - NOTE  | [FMDB/standard,FMDB/FTS] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sqlite3.h:8428:16: note: 'sqlite3_wal_checkpoint_v2' has been marked as being introduced in macOS 10.8 here, but the deployment target is macOS 10.6.0
>     - NOTE  | [FMDB/standard,FMDB/FTS] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:1340:15: note: enclose 'sqlite3_wal_checkpoint_v2' in an @available check to silence this warning
>     - WARN  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:310:49: warning: 'arc4random_uniform' is only available on macOS 10.7 or newer [-Wunguarded-availability]
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/stdlib.h:273:3: note: 'arc4random_uniform' has been marked as being introduced in macOS 10.7 here, but the deployment target is macOS 10.6.0
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  FMDB/src/fmdb/FMDatabase.m:310:49: note: enclose 'arc4random_uniform' in an @available check to silence this warning
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Execution policy exception registration failed and was skipped: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted" (in target 'FMDB' from project 'Pods')
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Execution policy exception registration failed and was skipped: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted" (in target 'Pods-App' from project 'Pods')
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  note: Execution policy exception registration failed and was skipped: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted" (in target 'App' from project 'App')
>     - NOTE  | [FMDB/standard, FMDB/FTS, FMDB/standalone, and more...] xcodebuild:  warning: Skipping code signing because the target does not have an Info.plist file and one is not being generated automatically. (in target 'App' from project 'App')
>     - WARN  | [FMDB/FTS,FMDB/standalone-fts] xcodebuild:  FMDB/src/extra/fts3/FMDatabase+FTS3.m:219:60: warning: 'NSPointerFunctionsWeakMemory' is only available on macOS 10.8 or newer [-Wunguarded-availability]
>     - NOTE  | [FMDB/FTS,FMDB/standalone-fts] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSPointerFunctions.h:40:5: note: 'NSPointerFunctionsWeakMemory' has been marked as being introduced in macOS 10.8 here, but the deployment target is macOS 10.6.0
>     - NOTE  | [FMDB/FTS,FMDB/standalone-fts] xcodebuild:  FMDB/src/extra/fts3/FMDatabase+FTS3.m:219:60: note: enclose 'NSPointerFunctionsWeakMemory' in an @available check to silence this warning
>     - NOTE  | [FMDB/standalone,FMDB/standalone-fts] xcodebuild:  note: Execution policy exception registration failed and was skipped: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted" (in target 'sqlite3' from project 'Pods')
>     - ERROR | [OSX] [FMDB/SQLCipher] unknown: Encountered an unknown error (CocoaPods could not find compatible versions for pod "SQLCipher":
>   In Podfile:
>     FMDB/SQLCipher (from `/Users/rryan/.../fmdb/FMDB.podspec`) was resolved to 2.7.7, which depends on
>       SQLCipher (~> 4.0)
> 
> Specs satisfying the `SQLCipher (~> 4.0)` dependency were found, but they required a higher minimum deployment target.) during validation.
> 
> [!] The spec did not pass validation, due to 1 error and 6 warnings.
> ```
> 
> Most of this consists of warnings that seem like they could be fixed by specifying the correct deployment target in the podspec. But the one that’s throwing me is the last one:
> 
> ```
>     - ERROR | [OSX] [FMDB/SQLCipher] unknown: Encountered an unknown error (CocoaPods could not find compatible versions for pod "SQLCipher":
>   In Podfile:
>     FMDB/SQLCipher (from `/Users/rryan/.../fmdb/FMDB.podspec`) was resolved to 2.7.7, which depends on
>       SQLCipher (~> 4.0)
> ```
> 
> I’m not quite sure what to make of that as it appears to be that `SQLCipher` does have a pod for v4.4, which seems like it should match. Does anyone know what’s going on there?
> 
> (By the way, I’m not sure why it’s using `~>` rather than, say, `>=`, but that’s a separate question.)

@robertmryan I don't think we ever tried re-publishing to CocoaPods trunk. Any chance you can help with this? Would love to access the latest release via CocoaPods trunk, rather than explicitly pointing my pod dependency to the HEAD of this repo's master branch.

**Comment by R4N** [comment]
@tstump-phunware 

>Specs satisfying the `SQLCipher (~> 4.0)` dependency were found, but they required a higher minimum deployment target.) during validation.

The latest podspec for SQLCipher specifies the minimum deployment targets of modern Xcode versions:

https://github.com/sqlcipher/sqlcipher/blob/v4.5.5/SQLCipher.podspec.json#L9

2.7.7 tag for FMDB doesn't include any minimum deploy target specifications in the podspec:

https://github.com/ccgus/fmdb/blob/2.7.7/FMDB.podspec

I had issued a PR and it was merged in to update the minimum deployment target of the FMDB podspec to match that in SQLCipher (min for modern versions of Xcode):

https://github.com/ccgus/fmdb/pull/849

>I’m not quite sure what to make of that as it appears to be that SQLCipher does have a pod for v4.4, which seems like it should match. Does anyone know what’s going on there?
(By the way, I’m not sure why it’s using ~> rather than, say, >=, but that’s a separate question.)

The `~>` will attempt to use the latest available version of SQLCipher subspec up to but not including the next major version (i.e 4.x.x but not 5.x.x). The main reason for this (instead of `>=`) is that  frequently major version releases of SQLCipher may/can change encryption parameters and won't work/open databases that were encrypted with the previous major version without an additional call to [PRAGMA cipher_migrate](https://www.zetetic.net/sqlcipher/sqlcipher-api/#cipher_migrate). Restricting the subspec to not exceed the next major SQLCipher version will allow a coordinated/intentional bump to the next major version of SQLCipher subspec (when it's available) so folks won't be upgraded to the next major version when updating the FMDB/SQLCipher pod unintentionally.

In summary:

* The latest available version of SQLCipher (4.x.x) subspec will attempt to be used (4.5.5 currently)
* 2.7.7 FMDB podspec doesn't include minimum deployment targets which are >= the min deployment targets of the SQLCipher subspec
* Utilizing the FMDB podspec which includes the min deployment target adjustment should allow the minimum deployment targets to match up and succeed
